### PR TITLE
TCK tests for Java SE Bootstrap API (and other post-3.0 additions)

### PR DESCRIPTION
**Pull request by mkarg**
This PR is a **draft** for a new TCK **only** covering post-3.0 additions to the Jakarta REST specifiction.

In particular it currently **only** contains **some** tests as a proof-of-concept. Hence it is neither complete nor intended to be actually merged *as-is*. Certainly this PR MAY be used as a fallback in case no *better* solution is found until deadline.

My proposal is to keep the original Platform TCK as-is, and use this PR as a *starting-point* for a second TCK with anything post-3.0, but not to import anything from the Platform TCK into this new TCK *for some time*, so we can decide *later* whether we live forever with two TCKs or whether we rebuild the existing tests in the new TCK. This scenario allows *every* contributor to write new tests **at once** without *any* learning curve, while it does not add *too much* additional work for users running this addtional test kit.

Please discuss, but keep in mind: This is not a wishlist to command *others*. Whather you want, YOU will be the one who has to deliver it in the next weeks! ;-)

### CODE REVIEW

File: jaxrs-tck/pom.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        <dependency>
+            <groupId>jakarta.ws.rs</groupId>
+            <artifactId>jakarta.ws.rs-api</artifactId>
+            <version>3.1-SNAPSHOT</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>[5.5.2, 5.6-A00)</version>
```
Comment (jansupol) [inline jaxrs-tck/pom.xml:91]: For these types of dependencies, the TCK bundle should file CQs, I believe. But the CQ can only be filed for a certain version, not for a range.

Comment (arjantijms) [reply jaxrs-tck/pom.xml:91]: Just wondering, shouldn't we create a tool that extracts the dependencies from Maven and automatically files CQs?

In the case of a range, the tool could check Maven central, check for every available version in that range, and file CQs for each and every one fo them?

Comment (mkarg) [reply jaxrs-tck/pom.xml:91]: @jansupol AFAIK ranges are allowed as long as the changes cover only bug fixes, but I will ask the EMO about it.

@arjantijms Sounds cool - do you have such a tool? Even better: This should be a bot on Github! ;-)

Comment (mkarg) [reply jaxrs-tck/pom.xml:91]: I resolved the following answer from EF's @waynebeaton:
>Hi Markus.
>
>CQs are not required for service releases (bug fixes only) based on major or minor versions of content that has already been approved. 
>
>I don't tend to see too many version ranges in Maven dependencies, but there's nothing stopping you from using them (except, perhaps, the ire of build purists). The way that I tend to deal with dependencies when doing an IP Log check is to check project builds to see that what actually gets resolved by Maven has been vetted by the process.
>
>We're taking steps to provide a tool that provide some additional automation of the vetting process. The Dash License Tool is pushing in that direction. There's a pull request to turn it into a Maven plug-in that I'll accept shortly. You'll also notice some experimenting with review request automation with the "-review" switch.

Comment (mkarg) [reply jaxrs-tck/pom.xml:91]: CQs have been approved by PMC meanwhile. :-)

**Comment by jansupol** [comment]
One thing I really miss with the new Arquillian based TCKs are the assertion IDs. Every original TCK test has an assertion ID, which is stored in the [list of all assertions](https://github.com/eclipse-ee4j/jakartaee-tck/tree/master/install/jaxrs/docs/assertions). It allows for immediate check what javadoc/spec part the test refers to.

The HTML assertion list are generated from the [XML files](https://github.com/eclipse-ee4j/jakartaee-tck/tree/master/internal/docs/jaxrs), which are filled either manually (spec) or automatically (javadoc) by [CTS tools](https://github.com/eclipse-ee4j/jakartaee-tck-tools).

**Comment by chkal** [comment]
@jansupol 

We have assertion IDs in the Arquillian-based Jakarta MVC TCK.

  * Assertion IDs are included in the AsciiDoc documents. See [here](https://github.com/eclipse-ee4j/mvc-api/blame/master/spec/src/main/asciidoc/chapters/data-binding.asciidoc#L84).
  * The build generates the XML file from these annotations. See [here](https://github.com/eclipse-ee4j/mvc-api/blob/e7381974d59b6ff426a3497ccd2b3ae6a8e3a892/spec/pom.xml#L122-L159).
  * The TCK contains annotations to reference the assertions. See [here](https://github.com/eclipse-ee4j/mvc-tck/blob/9c49c992e3e39bf742cd1c02d9d62885b9e29ddc/tests/src/main/java/jakarta/mvc/tck/tests/binding/types/BindingTypesTest.java#L69-L73).
  * The result is a corresponding coverage report. See [here](https://www.mvc-spec.org/spec/tck-coverage/index.html).

So it is definitely possible to do this.

**Comment by mkarg** [comment]
I do not see the relation of *Arquillian* to *this* PR, which *isn't* using Arquillian at all. Wrong thread, possibly?

**Review by spericas — DISMISSED** [review]
This looks like a great starting point. My only small comment is about the use of `final` in local variables. I would prefer to drop this and make the code less verbose.

**Comment by jansupol** [comment]
> I do not see the relation of _Arquillian_ to _this_ PR, which _isn't_ using Arquillian at all. Wrong thread, possibly?

No matter what test framework is used, there are still no assertion ids.

File: jaxrs-tck/src/main/java/jakarta/ws/rs/tck/uribuilder/UriBuilderIT.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * Compliance Test for URI Builder API of Jakarta REST API
+ *
+ * @author Markus KARG (markus@headcrashing.eu)
+ * @since 3.1
+ */
+@Timeout(value = 1, unit = HOURS)
+public final class UriBuilderIT {
+
+    /**
+     * Verifies that a valid can be created from scratch.
```
Comment (jansupol) [inline jaxrs-tck/src/main/java/jakarta/ws/rs/tck/uribuilder/UriBuilderIT.java:44]: A valid instance perhaps?

File: jersey-tck/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Jakarta REST TCK
+
+The **Jakarta REST TCK** is a standalone kit for testing compliance of an implementation with the Jakarta REST specification.
+
+**Note:** This TCK contains *only* tests being added since Jakarta REST 3.0. To proof compliance as part of a certification review, additional test have to be performed. Those are part of the Jakarta Platform TCK for historic reasons.
+
+
+## Performing Test
+
+While the test kit *is not dependent* of Maven, the most easy way to perform the tests is to create a copy of the sample Maven project found in the `jersey-tck` folder and adjust it to the actual needs of the implementation to be actually tested:
```
Comment (jansupol) [inline jersey-tck/README.md:10]: dependent *on* Maven?

**Review by jansupol — COMMENTED** [review]
This PR introduces a naming notation of `*IT` classes. What does it stand for? Is `IT` better then `Test`?

**Comment by mkarg** [comment]
> This PR introduces a naming notation of `*IT` classes. What does it stand for? Is `IT` better then `Test`?

This is Maven's default to separate integration tests (executed by `maven-failsafe-plugin`) from unit tests (executed by `maven-surefire-plugin`).

**Comment by mkarg** [comment]
> No matter what test framework is used, there are still no assertion ids.

Nobody so far sent me a link to an official public document of the WG's actual requirements on TCKs, so I keep things as simple as essentially possible *for now*. As I personally do not see a need for the assertion IDs, *I* will not invest time in that *for the time being*. So if somebody wants to have assertion IDs, feel free to open a PR in my fork, and I will be happy to adopt the commit containing them.

**Comment by mkarg** [comment]
> This looks like a great starting point. My only small comment is about the use of `final` in local variables. I would prefer to drop this and make the code less verbose.

This is personal style, and mine is to keep the finals because this is what Java is like and I definitively *want* constants. So unless there will be `const` in Java, let's agree not to dicuss style as it would lead to nowhere. ;-)

File: jaxrs-tck/src/main/java/jakarta/ws/rs/tck/uribuilder/UriBuilderIT.java (hunk 2 above)
Comment (mkarg) [reply jaxrs-tck/src/main/java/jakarta/ws/rs/tck/uribuilder/UriBuilderIT.java]: Good catch! Just fixed the typos.

**Comment by mkarg** [comment]
Guys, to get things forward I propose you accept the PR as-is unless it is definitively WRONG, and put your wanted changes ontop on your own. Otherwise there will not be progress.

**Comment by mkarg** [comment]
It is really sad to see that this thread is dead since weeks. In an attempt to get towards 3.1 with *minimal* effort, please let's just accept this PR as-is, *including* the force that implementors MUST execute two separate test sets, even if some of you *think* that would be forbidden. It definitively *is not*. It is up to *us* to decide about this, neither the steering committee nor the PMC so far prevented us from that, so please, let's move on basing on the actualy fact (not *myths*) and existing code (not code *to write*), not further delaying progress. It might be the  case that some of us personally don't like the idea of having two TCKs, but then *they* must provide the work to synchronize frequently; it is not an excuse to further suspend *this* PR.

**Comment by scottmarlow** [comment]
It is not dead to me nor is your platform tck issue for creating separate maven projects.

**Comment by spericas** [comment]
Hi @scottmarlow. What is your take on specs creating their own TCKs? Do you have any suggestions on how to do this?

**Comment by scottmarlow** [comment]
> Hi @scottmarlow. What is your take on specs creating their own TCKs? Do you have any suggestions on how to do this?

Hi @spericas.  Hopefully I will be able to respond soon and very much want to but am still focused on Jakarta EE 9.1 and trying to figure some things out for EE 9.1 still.

**Comment by spericas** [comment]
@mkarg I think we should push your changes to get the ball rolling here with these new TCKs. It's been a while, do you have any other changes or is this PR ready for review?

**Comment by mkarg** [comment]
Yes it's been quite a while, but IIRC I just kept it in draft state to separate the phase where we all discuss the idea in general from the actual voting phase. So as nobody else requested changes, I'm fine with diving right into the voting phase right now.

**Comment by andymc12** [comment]
I like the idea of using JUnit 5, and of creating a separate TCK artifact outside of the main CTS/TCK project. Thanks for getting this started. 

For other tests (like Multipart), I think we'll need to add dependencies on things like Arquillian and ShrinkWrap.  I can help with that as part of the Multipart TCK work. 

@mkarg, it's certainly been a while on this PR, since the copyrights still say 2020. :-) I think they need to say 2021, but then I'm ok with approving.

**Comment by spericas** [comment]
@andymc12 Good point, we should update the copyrights.

**Comment by mkarg** [comment]
In fact the copyright must stay as it is as it has to reflect the day of writing not the day of merging.

**Comment by mkarg** [comment]
Arquillian etc. should be added *when needed*. As *this* PR does not need it, feel free to add a second PR ontop.

**Comment by mkarg** [comment]
@chkal @arjantijms et al. Can you please cast your vote, so we can finally merge this?

**Review by chkal — APPROVED** [review]
Sorry for the delay.

Generally this looks good and is a great first step for our "own" TCK. Thanks to @mkarg for your work on this.

One minor note: IMO we will definitely need Arquillian in the future, but as mentioned before, this could be done AFTER merging this one. However, IIRC Arquillian doesn't support JUnit 5 yet!? However, we can address this later.

So +1 for merging!

**Review by andymc12 — APPROVED** [review]

**Pull request merged**